File: app/src/main/java/com/dpis/module/templates/TemplateConfigValueAdapters.java
package com.dpis.module.templates;

import com.dpis.module.viewport.ViewportTargetSpec;
import com.dpis.module.viewport.ViewportTargetType;

import com.dpis.module.*;


public final class TemplateConfigValueAdapters {
    private TemplateConfigValueAdapters() {
    }

    public static TemplateConfigValue fromViewportTargetSpec(
            ViewportTargetSpec viewportTargetSpec,
            String viewportApplyMode,
            Integer fontScalePercent,
            String fontApplyMode,
            String typefaceId,
            String fontHookDomainsRaw) {
        return fromViewportTargetSpec(
                viewportTargetSpec,
                null,
                null,
                null,
                viewportApplyMode,
                fontScalePercent,
                fontApplyMode,
                typefaceId,
                fontHookDomainsRaw);
    }

    public static TemplateConfigValue fromViewportTargetSpec(
            ViewportTargetSpec viewportTargetSpec,
            String viewportTargetType,
            String viewportApplyMode,
            Integer fontScalePercent,
            String fontApplyMode,
            String typefaceId,
            String fontHookDomainsRaw) {
        return fromViewportTargetSpec(
                viewportTargetSpec,
                viewportTargetType,
                null,
                null,
                viewportApplyMode,
                fontScalePercent,
                fontApplyMode,
                typefaceId,
                fontHookDomainsRaw);
    }

    public static TemplateConfigValue fromViewportTargetSpec(
            ViewportTargetSpec viewportTargetSpec,
            String viewportTargetType,
            Integer viewportScaleMilliPercentDraft,
            Integer viewportWidthDpDraft,
            String viewportApplyMode,
            Integer fontScalePercent,
            String fontApplyMode,
            String typefaceId,
            String fontHookDomainsRaw) {
        ViewportTargetSpec spec = viewportTargetSpec != null
                ? viewportTargetSpec
                : ViewportTargetSpec.off();
        String targetType = spec.isEnabled()
                ? spec.type()
                : TemplateConfigValue.normalizeViewportTargetType(viewportTargetType);
        Integer scaleMilliPercent = spec.isRelativeScale()
                ? spec.scaleMilliPercent()
                : null;
        Integer widthDp = spec.isAbsoluteDp()
                ? spec.absoluteWidthDp()
                : null;
        return new TemplateConfigValue(
                targetType,
                scaleMilliPercent,
                widthDp,
                viewportScaleMilliPercentDraft,
                viewportWidthDpDraft,
                viewportApplyMode,
                fontScalePercent,
                fontApplyMode,
                typefaceId,
                fontHookDomainsRaw);
    }

    public static ViewportTargetSpec toViewportTargetSpec(TemplateConfigValue value) {
        TemplateConfigValue normalized = value != null ? value : TemplateConfigValue.EMPTY;
        if (normalized.isRelativeScaleViewport()) {
            return ViewportTargetSpec.relativeScale(normalized.viewportScaleMilliPercent);
        }
        if (normalized.isAbsoluteDpViewport()) {
            return ViewportTargetSpec.absoluteDp(normalized.viewportWidthDp);
        }
        return ViewportTargetSpec.off();
    }

    public static PackageConfigValue toPackageConfigValue(TemplateConfigValue value) {
        TemplateConfigValue normalized = value != null ? value : TemplateConfigValue.EMPTY;
        return new PackageConfigValue(
                toViewportTargetSpec(normalized),
                normalized.viewportTargetType,
                normalized.viewportApplyMode,
                normalized.fontScalePercent,
                normalized.fontApplyMode,
                normalized.typefaceId,
                normalized.fontHookDomainsRaw,
                null,
                null);
    }
}

File: app/src/main/java/com/dpis/module/templates/TemplateConfigValueAdapters.kt
package com.dpis.module.templates

import com.dpis.module.config.PackageConfigValue
import com.dpis.module.viewport.ViewportTargetSpec

object TemplateConfigValueAdapters {
    @JvmStatic
    fun fromViewportTargetSpec(
        viewportTargetSpec: ViewportTargetSpec?,
        viewportApplyMode: String?,
        fontScalePercent: Int?,
        fontApplyMode: String?,
        typefaceId: String?,
        fontHookDomainsRaw: String?
    ): TemplateConfigValue {
        return fromViewportTargetSpec(
            viewportTargetSpec,
            null,
            null,
            null,
            viewportApplyMode,
            fontScalePercent,
            fontApplyMode,
            typefaceId,
            fontHookDomainsRaw
        )
    }

    @JvmStatic
    fun fromViewportTargetSpec(
        viewportTargetSpec: ViewportTargetSpec?,
        viewportTargetType: String?,
        viewportApplyMode: String?,
        fontScalePercent: Int?,
        fontApplyMode: String?,
        typefaceId: String?,
        fontHookDomainsRaw: String?
    ): TemplateConfigValue {
        return fromViewportTargetSpec(
            viewportTargetSpec,
            viewportTargetType,
            null,
            null,
            viewportApplyMode,
            fontScalePercent,
            fontApplyMode,
            typefaceId,
            fontHookDomainsRaw
        )
    }

    @JvmStatic
    fun fromViewportTargetSpec(
        viewportTargetSpec: ViewportTargetSpec?,
        viewportTargetType: String?,
        viewportScaleMilliPercentDraft: Int?,
        viewportWidthDpDraft: Int?,
        viewportApplyMode: String?,
        fontScalePercent: Int?,
        fontApplyMode: String?,
        typefaceId: String?,
        fontHookDomainsRaw: String?
    ): TemplateConfigValue {
        val spec = if (viewportTargetSpec != null)
            viewportTargetSpec
        else
            ViewportTargetSpec.off()
        val targetType = if (spec.isEnabled)
            spec.type()
        else
            TemplateConfigValue.normalizeViewportTargetType(viewportTargetType)
        val scaleMilliPercent = if (spec.isRelativeScale)
            spec.scaleMilliPercent()
        else
            null
        val widthDp = if (spec.isAbsoluteDp)
            spec.absoluteWidthDp()
        else
            null
        return TemplateConfigValue(
            targetType,
            scaleMilliPercent,
            widthDp,
            viewportScaleMilliPercentDraft,
            viewportWidthDpDraft,
            viewportApplyMode,
            fontScalePercent,
            fontApplyMode,
            typefaceId,
            fontHookDomainsRaw
        )
    }

    @JvmStatic
    fun toViewportTargetSpec(value: TemplateConfigValue?): ViewportTargetSpec {
        val normalized = if (value != null) value else TemplateConfigValue.EMPTY
        if (normalized == null) {
            return ViewportTargetSpec.off()
        }
        if (normalized.isRelativeScaleViewport) {
            return ViewportTargetSpec.relativeScale(normalized.viewportScaleMilliPercent)
        }
        if (normalized.isAbsoluteDpViewport) {
            return ViewportTargetSpec.absoluteDp(normalized.viewportWidthDp)
        }
        return ViewportTargetSpec.off()
    }

    @JvmStatic
    fun toPackageConfigValue(value: TemplateConfigValue?): PackageConfigValue {
        val normalized = if (value != null) value else TemplateConfigValue.EMPTY
        return PackageConfigValue(
            toViewportTargetSpec(normalized),
            normalized.viewportTargetType,
            normalized.viewportApplyMode,
            normalized.fontScalePercent,
            normalized.fontApplyMode,
            normalized.typefaceId,
            normalized.fontHookDomainsRaw,
            null,
            null
        )
    }
}
